src/mds/query.py

```diff
@@ -64,22 +72,388 @@ async def search_metadata(
+    objects having an authz string matching the pattern
+    "/programs/%/projects/%" at any index in its "_resource_paths" array? (%
+    has been used as the wildcard so far because that's what Postgres uses as
+    the wildcard for LIKE) In this case, the "1" and "3" objects should be
+    returned.
+
+    The filter syntax that was arrived at ending up following the syntax
+    specified by a Node JS implementation
+    (https://www.npmjs.com/package/json-api#filtering) of the JSON:API
+    specification (https://jsonapi.org/).
+
+    The format for this syntax is filter=(field_name,operator,value), in which
+    the field_name is a json key without quotes, operator is one of :eq, :ne,
+    :gt, :gte, :lt, :lte, :like, :all, :any (see operators dict), and value is
+    a typed json value against which the operator is run.
+
+    Examples:
+
+        - GET /objects?filter=(message,:eq,"morning") returns "2"
+        - GET /objects?filter=(counts.1,:eq,3) returns "3"
+
+    Compound expressions are supported:
+
+        - GET /objects?filter=(_resource_paths,:any,(,:like,"/programs/%/projects/%"))
+          returns "1" and "3"
+        - GET /objects?filter=(counts,:all,(,:eq,42))
+          returns "2"
+
+    Boolean expressions are also supported:
+
+        - GET /objects?filter=(or,(_uploader_id,:eq,"101"),(_uploader_id,:eq,"102"))
+          returns "1" and "2"
+        - GET /objects?filter=(or,(and,(pet,:eq,"ferret"),(sport,:eq,"soccer")),(message,:eq,"hello"))
+          returns "0", "1", and "2"
+
+    """
+    limit = min(limit, 2000)
+
+    def add_filter(target_key, operator_name, other):
+        """
+        XXX need a better name for other variable. with a scalar filter, other
+        is the value that the target value is compared against
+
+        e.g. for the following requests, the arguments to add_filter:
+            * `GET /objects?filter=(_uploader_id,:eq,"57")`
+                - target_key is 'uploader_id'
+                - operator_name is ':eq'
+                - other is '57'
+            * `GET /objects?filter=(count,:gte,16)`
+                - target_key is 'count'
+                - operator_name is ':gte'
+                - other is 16 (note this is a python int)
+            * `GET /objects?filter=(_resource_paths,:any,(,:like,"/programs/foo/%"))`
+                - target_key is '_resource_paths'
+                - operator_name is ':any'
+                - other is {'name': '', 'op': ':like', 'val': '/programs/foo/%'}
+        """
+
+        json_object = Metadata.data[list(target_key.split("."))]
+
+        if type(other) is dict:
+            #  since other isn't a primitive type, it means that this is a compound
+            #  operator
+            #  (e.g. GET objects?filter=(_resource_paths,:any,(,:like,"/programs/%")) )
+            return operators[operator_name]["sql_clause"](
+                json_object, other["op"], other["val"]
```

Comment:
if the filter stays user-provided, we might want to add a syntax validation step somewhere

Reply:
looks like `parsimonious` does that for us